source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -152,74 +173,80 @@ static int8_t ns_block_validate(int *block_start, int direction)
 static void *ns_dyn_mem_internal_alloc(const int16_t alloc_size, int direction)
 {
 #ifndef STANDARD_MALLOC
-    void *retval = 0;
+    int *retval = NULL;
     int data_size = convert_allocation_size(alloc_size);
+    int *hole_ptr = NULL;
+    int *ptr = NULL;
     if (data_size) {
-        int h_size = (heap_size / sizeof(int));
-        int *ptr = direction > 0 ? heap_main : heap_main_end;
-        int moved = 0;
-        bool update_start = true;
         platform_enter_critical();
-        while (moved < h_size) {
+        //use ns_list_foreach_reverse if direction == -1;
+        hole_t *cur_hole;
+        if ( direction > 0 ) {
+            cur_hole = ns_list_get_first(&holes_list);
+        } else {
+            cur_hole = ns_list_get_last(&holes_list);
+        }
+
+        while ( cur_hole ) {
+            ptr = block_start_from_hole(cur_hole);
             if (ns_block_validate(ptr, direction) == 0) {
-                int block_data_size = *ptr;
-
-                if (block_data_size < 0) {
-                    //This is a free block so stop updating pointer
-                    update_start = false;
-
-                    block_data_size = -block_data_size;
-                    if (block_data_size >= data_size) {
-                        /*found block*/
-                        if (block_data_size > (data_size + 4)) {
-                            //There is enough room for a new hole so create it first
-                            int *hole_ptr;
-                            if (direction > 0) {
-                                hole_ptr = ptr + data_size + 2;
-                            } else {
-                                hole_ptr = ptr - (data_size + 2);
-                            }
-                            int hole_size = block_data_size - data_size - 2;
-                            *hole_ptr = -(hole_size);
-                            if (direction > 0) {
-                                hole_ptr += (hole_size + 1);
+                int block_data_size = -(*ptr);
```

Comment:
The validation call is checking it's a valid block, either free or allocated. In this case we should be specifically checking it's free.